packages/cojson-transport-ws/src/index.ts

```diff
@@ -114,23 +81,114 @@ export function createWebSocketPeer({
+        setBatchingEnabled(enabled: boolean) {
+            batchingEnabled = enabled;
+        },
+        close() {
+            outgoingMessages.close();
+        },
+    };
+}
+
+function createClosedEventEmitter(callback = () => {}) {
+    let disconnected = false;
+
+    return () => {
+        if (disconnected) return;
+        disconnected = true;
+        callback();
+    }
+}
+
+export function createWebSocketPeer({
+    id,
+    websocket,
+    role,
+    expectPings = true,
+    batchingByDefault = true,
+    onClose,
+}: CreateWebSocketPeerOpts): Peer {
+    const incoming = new cojsonInternals.Channel<
+        SyncMessage | DisconnectedError | PingTimeoutError
+    >();
+    const emitClosedEvent = createClosedEventEmitter(onClose);
+
+    function handleClose() {
+        incoming
+            .push("Disconnected")
+            .catch((e) =>
+                console.error("Error while pushing disconnect msg", e),
+            );
+        emitClosedEvent();
+    }
+
+    websocket.addEventListener("close", handleClose);
+
+    const pingTimeout = createPingTimeoutListener(expectPings, () => {  
+        incoming
+            .push("PingTimeout")
+            .catch((e) => console.error("Error while pushing ping timeout", e));
+        emitClosedEvent();
+    });
+
+    const outgoingMessages = createOutgoingMessagesManager(websocket, batchingByDefault);
+
+    function handleIncomingMsg(event: { data: unknown }) {
+        const result = deserializeMessages(event.data);
+
+        if (!result.ok) {
+            console.error(
+                "Error while deserializing messages",
+                event.data,
+                result.error,
+            );
+            return;
+        }
+
+        const { messages } = result;
+
+        if (messages.length > 1) {
+            // If more than one message is received, the other peer supports batching
+            outgoingMessages.setBatchingEnabled(true);
+        }
+
+        pingTimeout.reset();
+
+        for (const msg of messages) {
+            if (msg && "action" in msg) {
+                incoming
+                    .push(msg)
+                    .catch((e) =>
+                        console.error("Error while pushing incoming msg", e),
+                    );
+            }
+        }
+    }
+
+    websocket.addEventListener("message", handleIncomingMsg);
+
     return {
         id,
         incoming,
         outgoing: {
-            push: pushMessage,
+            push: outgoingMessages.sendMessage,
             close() {
                 console.log("Trying to close", id, websocket.readyState);
-                if (supportsBatching) {
-                    outgoingMessages.close();
-                }
+                outgoingMessages.close();
+
+                websocket.removeEventListener("message", handleIncomingMsg);
+                websocket.removeEventListener("close", handleClose);
+                pingTimeout.clear();
+                emitClosedEvent();
```

Comment:
Refactored the close management to detach all the events and clear all the pending timeouts